### PR TITLE
Update Mimir Prometheus to `baf7468`

### PR DESCRIPTION
**Pull request by gotjosh**
#### What this PR does

The whole list of changes is included as part of Mimir Prometheus PR in https://github.com/grafana/mimir-prometheus/pull/622,  https://github.com/grafana/mimir-prometheus/pull/623 and https://github.com/grafana/mimir-prometheus/pull/625

~~I don't think there's any change that's worth mentioning, given that they aren't user-facing but do let me know if you think there's something worth adding to the changelog.~~

As part of the second `mimir-prometheus` PR there is a change that I have noted in the changelog. Matrix results were returning `null` instead of `[]` in some cases.

I have managed to reproduce this by running:

```bash
# Main branch
$ ~> curl "http://localhost:61793/prometheus/api/v1/query_range?query=bottomk(2,%20notExists)&start=1714745905&end=1714745910&step=30s" \
     -H 'X-Scope-OrgID: tenant-1' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    65  100    65    0     0  24780      0 --:--:-- --:--:-- --:--:-- 32500
{
  "status": "success",
  "data": {
    "resultType": "matrix",
    "result": null
  }
}


# This PR

$ ~> curl "http://localhost:61235/prometheus/api/v1/query_range?query=bottomk(2,%20notExists)&start=1714745710&end=1714745715&step=30s" \
     -H 'X-Scope-OrgID: tenant-1' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    63  100    63    0     0   6039      0 --:--:-- --:--:-- --:--:--  6300
{
  "status": "success",
  "data": {
    "resultType": "matrix",
    "result": []
  }
}


```

Against a querier in both `main` and this `branch`.

#### Which issue(s) this PR fixes or relates to

Related to https://github.com/grafana/mimir-squad/issues/1845

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Review by pracucci — APPROVED** [review]
LGTM. I've reviewed mimir-prometheus changes, and checked [protobuf release note](https://github.com/protocolbuffers/protobuf-go/releases/tag/v1.34.0) (I can't see anything suspicious there).

File: CHANGELOG.md [hunk 1]
```diff
@@ -33,7 +33,7 @@
 * [BUGFIX] Rules: improve error handling when querier is local to the ruler. #7567
 * [BUGFIX] Querier, store-gateway: Protect against panics raised during snappy encoding. #7520
 * [BUGFIX] Ingester: Prevent timely compaction of empty blocks. #7624
-* [BUGFIX] querier: Don't cache context.Canceled errors for bucket index. #7620
+* [BUGFIX] Querier: Don't cache context.Canceled errors for bucket index. #7620
```
Comment (gotjosh) [inline CHANGELOG.md:36]: I think this is a typo all the others are capitalised.

Comment (pracucci) [reply CHANGELOG.md:36]: It's OK editing it, but we're not very strict with the CHANGELOG. Anyway, it's good you improved it.

**Comment by gotjosh** [comment]
The CI failures are real; they relate to https://github.com/prometheus/prometheus/pull/14047, which is going to be shortly merged. 

What are my options here? I could technically fix it in mimir-Prometheus very quickly and re-import or just wait?

I'd like this PR to be merged today so that this gets rolled out next week 😭

**Comment by pstibrany** [comment]
> I'd like this PR to be merged today so that this gets rolled out next week 😭

It looks like that PR will indeed be merged shortly. I'd suggest to wait for it and if it's not merged by monday, do the fix in mimir-prometheus. We can update our weekly release after it's been cut.

**Comment by ArthurSens** [comment]
Looks like that PR is blocking your work, I'm just going to add the missing test case to the contributor's PR. Do you want to add a review over there? :)

**Review by johannaratliff — APPROVED** [review]

**Pull request merged**